sklearn/manifold/tests/test_t_sne.py

```diff
@@ -806,3 +814,14 @@ def test_tsne_with_different_distance_metrics():
             metric='precomputed', n_components=n_components_embedding,
             random_state=0).fit_transform(dist_func(X))
         assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
+
+
+def test_tsne_with_csr_distance_metrics():
+    """Make sure that TSNE works identically for sparse and dense matrix"""
+    dist = np.zeros((3, 3))
+    dist_csr = sp.csr_matrix(dist)
+    tsne = TSNE(metric="precomputed")
+    X_transformed_dense = tsne.fit_transform(dist)
```

Comment:
@jnothman  

> Check that tSNE results for sparse X are the same as for dense X where X is a feature matrix and where X is a precomputed distance matrix *with all zeros explicit*

Does this properly implement what you intended by this statement? I am not sure since the output of neighbors inside TNSE.fit give the same result. I guess this function needs modification.

Reply:
No, I don't mean that the entire matrix is zeros. I mean that the zero distances on the diagonally should be explicitly in X.data, as opposed to an element not explicitly stored (and hence assumed to have a value of zero).

You can construct an all-explicit CSR matrix with:
```
def array_to_explicit_csr(X):
    return csr_matrix((X.ravel(), np.tile(np.arange(X.shape[1]), X.shape[0]), np.arange(0, X.size + 1, X.shape[1])))
```